tajo-core/src/main/java/org/apache/tajo/master/rm/NodeStatus.java

```diff
@@ -0,0 +1,293 @@
+
+  /** Total resources on the node. */
+  private final NodeResource totalResourceCapability;
+
+  /** Node connection information */
+  private WorkerConnectionInfo connectionInfo;
+
+  private static final ReconnectNodeTransition RECONNECT_NODE_TRANSITION = new ReconnectNodeTransition();
+  private static final StatusUpdateTransition STATUS_UPDATE_TRANSITION = new StatusUpdateTransition();
+
+  private static final StateMachineFactory<NodeStatus,
+      NodeState,
+      NodeEventType,
+      NodeEvent> stateMachineFactory
+      = new StateMachineFactory<NodeStatus,
+      NodeState,
+      NodeEventType,
+      NodeEvent>(NodeState.NEW)
+
+      // Transition from NEW
+      .addTransition(NodeState.NEW, NodeState.RUNNING,
+          NodeEventType.STARTED,
+          new AddNodeTransition())
+
+      // Transition from RUNNING
+      .addTransition(NodeState.RUNNING, EnumSet.of(NodeState.RUNNING, NodeState.UNHEALTHY),
+          NodeEventType.STATE_UPDATE,
+          STATUS_UPDATE_TRANSITION)
+      .addTransition(NodeState.RUNNING, NodeState.LOST,
+          NodeEventType.EXPIRE,
+          new DeactivateNodeTransition(NodeState.LOST))
+      .addTransition(NodeState.RUNNING, NodeState.RUNNING,
+          NodeEventType.RECONNECTED,
+          RECONNECT_NODE_TRANSITION)
+
+      // Transitions from UNHEALTHY state
+      .addTransition(NodeState.UNHEALTHY, EnumSet.of(NodeState.RUNNING, NodeState.UNHEALTHY),
+          NodeEventType.STATE_UPDATE,
+          STATUS_UPDATE_TRANSITION)
+      .addTransition(NodeState.UNHEALTHY, NodeState.LOST,
+          NodeEventType.EXPIRE,
+          new DeactivateNodeTransition(NodeState.LOST))
+      .addTransition(NodeState.UNHEALTHY, NodeState.UNHEALTHY,
+          NodeEventType.RECONNECTED,
+          RECONNECT_NODE_TRANSITION);
+
+  private final StateMachine<NodeState, NodeEventType, NodeEvent> stateMachine =
+      stateMachineFactory.make(this, NodeState.NEW);
+
+  public NodeStatus(TajoRMContext rmContext, NodeResource totalResourceCapability, WorkerConnectionInfo connectionInfo) {
+    this.rmContext = rmContext;
+
+    this.connectionInfo = connectionInfo;
+    this.lastHeartbeatTime = System.currentTimeMillis();
+    this.totalResourceCapability = totalResourceCapability;
+    this.availableResource = NodeResources.clone(totalResourceCapability);
+  }
+
+  public int getWorkerId() {
+    return connectionInfo.getId();
+  }
+
+  public WorkerConnectionInfo getConnectionInfo() {
+    return connectionInfo;
+  }
+
+  public void setLastHeartbeatTime(long lastheartbeatReportTime) {
+    HEARTBEAT_TIME_UPDATER.lazySet(this, lastheartbeatReportTime);
+  }
+
+  public void setNumRunningQueryMaster(int numRunningQueryMaster) {
+    RUNNING_QM_UPDATER.lazySet(this, numRunningQueryMaster);
+  }
+
+  public int getNumRunningQueryMaster() {
+    return numRunningQueryMaster;
+  }
+
+  public void setNumRunningTasks(int numRunningTasks) {
+    RUNNING_TASK_UPDATER.lazySet(this, numRunningTasks);
```

Comment:
``RUNNING_TASK_UPDATE`` contains the number of running tasks. But, the getter method ``getNumRunningTasks()`` uses just the member variable ``numrunningTasks``. It seems to be invalid.